### ISSUE TITLE
MacOS GUI issues with V8.0.1

### ISSUE THREAD

**Issue by ridgeworks** [SWI-Prolog/issues#83]
SWI-Prolog version 8.0.1 on MacOS 10.13.4

The Mac GUI has regressed since v7.6.4:

1. In general the graphics interface is low-resolution. This applies to the fonts in the top-level window as well as any popup menus.

2. The font cannot be changed using the main menu item Settings/Font ... 

3. Application crashes on main menu item Settings/GUI preferences ...

**Comment by ridgeworks** [comment]
So I'm observing this on MacOS 10.13 and I strongly suspect this is an OS dependant issue. I do have Xquartz installed and none of the reported behaviour occurs using 7.6.4.

I'm talking about the main Prolog window (top level query interface). Also any system pop-up menus (file selection, font selection, etc.) are also "low-res" so it's probably some kind of application level setting.

**Comment by JanWielemaker** [comment]
I'm running MacOS 10.11.  The machine is too old to upgrade according to Apple.   I'll attend the mailinglist about this issue.   See who is affected ...

**Comment by ridgeworks** [comment]
It may also require a Retina class display to see the difference. I've attached a couple of screen dumps. It's a bit hard to distinguish the two but it's much more apparent looking at the screen.

Are you saying the Settings/GUI Preferences... didn't crash? I can send you the crash log entry if it would be helpful.

**Comment by JanWielemaker** [comment]
> Retina class display

I don't have that.  Not needed for my eyes :)  

> didn't crash

Yip (works fine).  Well, we'll see whether there is response on the list.  I doubt I can do much for this as it doesn't reproduce and my C++/Qt knowledge is very limited.

**Comment by ridgeworks** [comment]
Here's the start of the crash log. Appears to be where you suspected, i.e., not much to do with SWIP.

Process:               SWI-Prolog [44861]
Path:                  /Applications/SWI-Prolog.app/Contents/MacOS/SWI-Prolog
Identifier:            org.swi-prolog.app
Version:               ??? (???)
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           SWI-Prolog [44861]
User ID:               502

Date/Time:             2019-01-30 11:14:04.115 -0500
OS Version:            Mac OS X 10.13.4 (17E199)
Report Version:        12
Anonymous UUID:        1A55B400-E477-58D9-CD8F-644CB5C355E4

Sleep/Wake UUID:       8A7A0DE9-8702-4391-9B27-EA0D2119B7D5

Time Awake Since Boot: 8500000 seconds
Time Since Wake:       8400000 seconds

System Integrity Protection: disabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_CRASH (SIGILL)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Illegal instruction: 4
Termination Reason:    Namespace SIGNAL, Code 0x4
Terminating Process:   SWI-Prolog [44861]

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff56d3d20a mach_msg_trap + 10
1   libsystem_kernel.dylib        	0x00007fff56d3c724 mach_msg + 60
2   com.apple.CoreFoundation      	0x00007fff2e844045 __CFRunLoopServiceMachPort + 341
3   com.apple.CoreFoundation      	0x00007fff2e843397 __CFRunLoopRun + 1783
4   com.apple.CoreFoundation      	0x00007fff2e842a07 CFRunLoopRunSpecific + 487
5   com.apple.HIToolbox           	0x00007fff2db20d96 RunCurrentEventLoopInMode + 286
6   com.apple.HIToolbox           	0x00007fff2db20b06 ReceiveNextEventCommon + 613
7   com.apple.HIToolbox           	0x00007fff2db20884 _BlockUntilNextEventMatchingListInModeWithFilter + 64
8   com.apple.AppKit              	0x00007fff2bdd3a73 _DPSNextEvent + 2085
9   com.apple.AppKit              	0x00007fff2c569e34 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 3044
10  com.apple.AppKit              	0x00007fff2bdc8885 -[NSApplication run] + 764
11  libqcocoa.dylib               	0x000000010c8de8ed 0x10c8b4000 + 174317
12  org.qt-project.QtCore         	0x000000010a77b3b2 QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) + 418
13  org.qt-project.QtCore         	0x000000010a77ff72 QCoreApplication::exec() + 402
14  org.swi-prolog.app            	0x000000010994e4e9 main + 137
15  libdyld.dylib                 	0x00007fff56bf6015 start + 1

**Comment by ridgeworks** [comment]
A few more observations:

1. Anything using XWindows seems to crash, e.g., `profile/1`, graphical debugger, menu item `File/Navigator...`

2. The `About SWI-Prolog` menu item appears to do nothing on first use, then hangs (requires a `Force-Quit...`) on second use.

All the above as expected in 7.6.4.

**Comment by ridgeworks** [comment]
Perhaps, but it is your release and I wouldn't classify it as "stable". Maybe there's someting amiss with my configuration, although the behaviour is exactly the same on my MacBook and I'll poll a few of my colleagues to see what but their experiece is. 

But there definitely seems to be a regression. If there's a new dependancy it should be documented. Somebody must have built this Mac binary and (hopefully) tested it. IMO it should be either fixed or pulled from circulation. Failing that, publish a list of known working configurations for MacOS.

I'm quite happy with 7.6.4 until something better comes along. I just don't want to waste time trying to use something that's buggy.

That said, I do very much appreciate all your (collective) efforts over the many years. I'm happy to spend some time beta testing if that would help but I'm afraid I'm not much use when it comes to building applications.

**Comment by JanWielemaker** [comment]
If it gets fixed I'm happy.  I do not consider the Qt console on MacOS a core component.  If you want to contribute you could try to figure out which version is the first to exhibit these issues.  Most of the changes happened around 7.7.20.  Using the _show all files_ link on the download page you can download all versions.   Sorry ...

**Comment by ridgeworks** [comment]
Yes. 7.7.19 works, and I mean everything - hi-res fonts, font settings, About.. etc.  7.7.20-1 is broken. 7.7.25-1 is also broken so I suspect everything with a '-1' suffix is broken.

A friend with a MacBook Pro running MacOS 10.14.2 captured this snapshot as it was crashing (using a QuickTime movie).

**Comment by ridgeworks** [comment]
Console log:


Last login: Wed Oct 17 09:55:38 on console
Mac-Pro:~ rworkman$ /Applications/SWI-Prolog.app/Contents/MacOS/swipl ; exit;
Welcome to SWI-Prolog (threaded, 64 bits, version 8.0.1)
SWI-Prolog comes with ABSOLUTELY NO WARRANTY. This is free software.
Please run ?- license. for legal details.

For online help and background, visit http://www.swi-prolog.org
For built-in help, use ?- help(Topic). or ?- apropos(Word).

?- emacs.
XPCE Xt warning: locale not supported by Xlib, locale set to C
XPCE Xt warning: X locale modifiers not supported, using default
Fontconfig error: Cannot load default config file
Illegal instruction: 4
logout
Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.
Deleting expired sessions...43 completed.

[Process completed]



Just an aside. It would be a good idea if the default font was fixed width, e.g., I always set it to "Courier New".

**Comment by JanWielemaker** [comment]
I'm one step further.  After `chmod 0 /opt/local` I get the same crash.  I guess that implies it gets something from MacPorts that is not in the bundle.

**Comment by JanWielemaker** [comment]
... But the so-called working version 7.7.19 also crashes the same way if I block access to `/opt/local`.  
Seems `dtruss` is broken on MacOS 10.11, which makes debugging rather hard.  I'm giving up for now.   Will ask around.  I think this is something that must be fixed by someone who knows and uses a Mac.

**Comment by ridgeworks** [comment]
I believe MacOS's SIP (System Integrity Protection) has to be disabled for dtruss to work. Can't remember what the state of that was on 10.11. Disabling it is a bit complicated and I don't expect you to spend effort on it. 

I can run dtruss on my desktop (SIP disabled so I can run newer OS on my antiquated hardware), but it doesn't produce anything for -n SWI-Prolog so not quite sure what's going on. Hope you can find somebody to assist but let me know if there's anything else I can do.

**Comment by JanWielemaker** [comment]
The Qt menus not functioning seems to be caused by the switch to the CMake building process as it reproduces on Linux.  Still, both issues are very hard to debug :(

**Comment by JanWielemaker** [comment]
That is better.  Now ideally both a good and bad one and as attachment.  It is a bit hard to handle from the web page and getting it out properly is hard too :(  Sorry to bother ...

**Comment by ridgeworks** [comment]
Sorry about that - it was an attachment on the email but I guess GitHub thought there was a better way. Here's the traces for 8.0.1 and 7.6.4. 

[dtruss801For83.txt](https://github.com/SWI-Prolog/issues/files/2822599/dtruss801For83.txt)

[dtruss764For83.txt](https://github.com/SWI-Prolog/issues/files/2822702/dtruss764For83.txt)

**Comment by JanWielemaker** [comment]
Please try the just released 8.1.1.  Somehow the _Font_ menu doesn't work for me, but then it doesn't work for the 7.7.19 version for me either and you claimed that version was fine, so that may be a local problem for me.  The _Font_ menu does work fine on Linux now, while that version was broken as well.  Menus that start xpce work fine, also after making `/opt/local` inaccessible.

**Comment by ridgeworks** [comment]
Testing 8.1.1:

1. The font issue remains. Low-res fonts are used and the main menu item Settings/Font... brings up the font selection dialog but has no effect. This is the same behaviour as 7.7.20-1 but not the case in 7.7.19 (and before). This is likely a MacOS only problem, although it sounds like you've fixed something on Linux.

2. Any X-Windows interactions (GUI Preferences... , Navigator... , profile/1, etc.) now seem to work fine.

3. The excessive garbage collection problem (#85) has been fixed and the heat exchanger problem runs even faster than before (~10-15%); much appreciated.

**Comment by JanWielemaker** [comment]
Thanks for testing.  At least it is usable again.  I'll check whether I can find something obvious, but this entire Qt and MacOS bundling is a bit of a black box to me ...

I think you'll gain some more performance if you get rid of (most of) the exceptions.  Exception processing is relatively expensive, despite the fact that SWI-Prolog is not bad at it compared to other Prolog systems.

**Comment by ridgeworks** [comment]
Yes, very usable. What's left is a cosmetic issue but a bit annoying because it is a regression.

Unfortunately when handling user arithmetic data it's impossible to entirely avoid exception handling. My testing indicates that while wrapping the arithmetic in a catch/3 incurs some expense, the big overhead hit is on recovery from the exception. So in my particular case it's a matter of communicating the expense of handling infinities to users. For this particular "benchmark" that happens less than 5% of the time but there are probably ways to reduce it at the application level.

Avoiding floating point exceptions would seem to involve testing all operands for infinities before evaluating which would probably negate much of any performance gain while considerably complicating the code and it would still not catch cases where inifinities weren't involved. 

But you make a valid point and I need to pay attention to these overheads. A bigger issue, and one I haven't really looked into yet, is the cost of garbage collection - this benchmark spends ~20% of the time doing gc, according to the profiler, and that doesn't seem to be atypical in many of my larger examples.

**Comment by JanWielemaker** [comment]
The failing font change is due to a [bug in Qt](https://bugreports.qt.io/browse/QTBUG-69878).
I tried to upgrade Qt, but the new version does not run on my old MacOS (10.11) which I cannot upgrade because the hardware is too old according to Apple.  So, I fear we're stuck until I get new hardware or someone else builds the binary.

**Comment by ridgeworks** [comment]
It appears that this a regression in Qt, i.e., according to that bug report, Qt 5.10 works and Qt 5.11 does not. This probably explains why SWIP pre 7.6.20 works and now is "broken". Is there a possibility that SWIP could go back to using 5.10 (or whatever 7.7.19 used) until the build situation gets sorted out? Or has that ship already sailed?

My other option is just to use swipl as a Terminal application. I have to admit this isn't as bad as I had imagined, although I do miss the top level menus, particularly 'Consult..'. and 'Reload modified files'.

The larger question is support for SWIP on MacOS. It appears that you aren't currently able to build and test on recent versions of OS and other required libraries, e.g., Qt, and no else has been tasked with this role. Either this should be addressed or some redefinition of the MacOS "product" is required.

Anyway, all has been explained, and that's fine for now.

**Comment by JanWielemaker** [comment]
> 'Consult..'. and 'Reload modified files'.

Know that after typing `?- ['..` you can hit _TAB_ to complete on file names.  You normally start Prolog from the terminal as `swipl file.pl` to start.  To reload the files, simply type `?- make.`.   I always use the terminal ...

The larger question on how to continue supporting this app has my attention.  I do plan to keep it alive, for now assuming the hardware will be replaced _by magic_ ...

**Comment by JanWielemaker** [comment]
>     1. In general the graphics interface is low-resolution. This applies to the fonts in the top-level window as well as any popup menus.

Seems to be caused by the plist app metadata generated by CMake.  See https://cmake.org/pipermail/cmake/2019-March/069152.html

My knowledge about Mac apps is really limited ...

**Comment by ridgeworks** [comment]
I probably don't know much more than you, but as a quick hack I saved the 8.0.2 info.plist file and overwrote it with the 7.6.4 version. On my laptop at  least, the low resolution fonts have been replaced with hi res versions. However, the font setting dialog still has no effect.

The two plist files are quite different; notably the 8.0.2 version is missing many settings (like the app version number). I don't have time right now to sort out which differences matter, but I'll dig into later next week and let you know what I discover (unless someone beats me to it).

**Comment by JanWielemaker** [comment]
Ok.  At least we know the problem now.  The font dialog not working as a Qt bug as we know.   I hope to get a more recent Mac at some point.  Until then my options are limited.  I refuse to buy a Mac on my own budget as, except for compiling and debugging SWI-Prolog I have no need for it.

**Comment by ridgeworks** [comment]
I don't think it's reasonable to ask you to personally invest either monetarily or intellectually in a platform that you're not going to use. But it would be helpful to identify somebody in the SWIP user community to take on the build role. Maybe you could list the set of requirements for somebody taking on this role and solicit help via the forum. I'm happy to participate but I it'll take a while for me to get up to speed. Other than that I'm not sure what to suggest.

As to the immediate problem, a minimal fix is described below with the caveat that I don't really understand the big picture from either a Mac or SWIP perspective.

As the link in your prior email suggested, the problem is caused by the missing NSPrincipalClass entry. I'm not sure how CMake generates (?) the plist file, but at a minimum the following two lines should be added to info.plist found at the MacOS package top level:

	<key>NSPrincipalClass</key>
	<string>NSApplication</string>

I would also suggest the short version number and bundle signature contain the appropriate values, e.g.:

	<key>CFBundleShortVersionString</key>
	<string>8.0.2</string>
	<key>CFBundleSignature</key>
	<string>swpl</string>

I've attached a copy of the file I've been using for testing and it seems to work fine. 

There are a number of other differences with the 7.6.4 version but it's not clear to me if they're significant or relevant to the new make process.

**Comment by JanWielemaker** [comment]
Thanks.  That should help.   Just as a sidenote, proper support for a platform requires me to have access.  It is quite common that people contact me with portability issues.  If they can't fix it themselves, mailing back and forth steps to try and results is too tedious.  These days most stuff runs on x64 and I can run most in a VM or even docker container.   That doesn't work for MacOS in a simple and legal way though and I refuse to run illegal software.

Note that Paulo takes care of the Macports version and I think the Homebrew version also has some support.

Ideally we have someone with good knowledge of the Mac eco system (app structure, bundles, frameworks, dynamic linking, etc), C and CMake.   That is quite a lot to ask.

**Comment by JanWielemaker** [comment]
Pushed SWI-Prolog/swipl-devel@f5970a6e077402745b2d7ea0805559f49f21dac6 to fix that.  Next release will be build on a new Mac using Qt 5.12.3, which also fixes the menu issues.  Added short version too.  The CFBundleSignature seems to be something you must register with Apple.

**Comment by ridgeworks** [comment]
Thanks, I look forward to trying it out.

From my rudimentary understanding, CFBundleSignature is mainly for compatibility with old Mac Classic apps and isn't used much anymore. I orginally mentioned it because pre-cmake versions of SWIP defined a value.

**Comment by ridgeworks** [comment]
v8.1.5 (assumed to be the Mac bundle incorporating this push) now *requires* MacOS 10.14. 

Is this the intent? Is there a framework/library dependancy that mandates this?

**Comment by JanWielemaker** [comment]
Preferably not of course :(  What happens?  On https://doc-snapshots.qt.io/qt5-5.12/macos.html it is claimed  that Qt 5.12 requires 10.12 or later.   I think it is claimed that without any special actions it will configure support for all versions as of 10.12.  In our case we do not use qmake but cmake for the configuration, so it is not impossible that something is needed.  I'd first like to know whether this is a Qt thing though.

Do the comand line tools (e.g., Contents/MacOS/swipl) work properly?

**Comment by JanWielemaker** [comment]
Uploaded https://eu.swi-prolog.org/download/devel/bin/swipl-8.1.5-2.x86_64.dmg, compiled with `MACOSX_DEPLOYMENT_TARGET=10.12`.  No clue whether or not it helps.

**Comment by ridgeworks** [comment]
I'm currently running macOS 10.13.6. 

With SWIP v8.1.15-1:

1. The MacOS bundle icon when viewed with the Finder has a "circle/slash" (⦸) overlay indicating it can't be used. When it is launched, a OS dialog comes up: "

**`You can’t use this version of the application “SWI-Prolog.app” with this version of macOS.`**
`You have macOS 10.13.6. The application requires macOS 10.14 or later.`

It was my impression that the minimum system version was controlled by the LSMinimumSystemVersion property in the Info.plist file but that's not the case here, so there must be some other mechanism that's general enough for the Finder to see it.


2. When I try execute the command line version, I get different results:

On my MacBook-10.13.6, it seems to be fine with the normal SWIP lead-in and prompt. (It complains that it "Cannot read termcap database" but I assume that's something I haven't configured properly on the MacBook.)

On my old MacPro-10.13.6, I get a trap:

`$ /Applications/SWI-Prolog.app/Contents/MacOS/swipl

SWI-Prolog [thread 1 () at Mon Apr 22 09:59:15 2019]: received fatal signal 4 (ill)
C-stack trace labeled "crash":
  [0] /Applications/SWI-Prolog.app/Contents/Frameworks/libswipl.8.1.5.dylib(save_backtrace+0xee) [0x101eab95e]
  [1] /Applications/SWI-Prolog.app/Contents/Frameworks/libswipl.8.1.5.dylib(sigCrashHandler+0xb2) [0x101eabf82]
  [2] /Applications/SWI-Prolog.app/Contents/Frameworks/libswipl.8.1.5.dylib(dispatch_signal+0x233) [0x101e381d3]
  [3] /usr/lib/system/libsystem_platform.dylib(_sigtramp+0x1a) [0x7fff64fe6f5a]
  [5] /Applications/SWI-Prolog.app/Contents/Frameworks/libswipl.8.1.5.dylib(setupProlog+0x4fc) [0x101e373ac]
  [6] /Applications/SWI-Prolog.app/Contents/Frameworks/libswipl.8.1.5.dylib(PL_initialise+0x1194) [0x101e61644]
  [7] /Applications/SWI-Prolog.app/Contents/MacOS/swipl(main+0x9) [0x101d99f49]
  [8] /usr/lib/system/libdyld.dylib(start+0x1) [0x7fff64cd8015]
Running on_halt hooks with status 132
Killing 3751 with default signal handlers`

v8.1.14 works fine on both Mac's, so I don't know how to explain the difference.


I just tested "swipl-8.1.15-2".

The bundle now works fine on the MacBook, but I get the same trap on the MacPro. I've attached a partial MacOS crash report for more info. 

[MacPro8.15Trap.txt](https://github.com/SWI-Prolog/issues/files/3103780/MacPro8.15Trap.txt)

In summary, not sure what to make of all this. Doesn't seem to be a QT issue (command line tool also traps) but specific to my (old) MacPro configuration and 8.1.15.

**Comment by JanWielemaker** [comment]
Linked from https://swi-prolog.discourse.group/t/macos-bundle/569.  So far no response.

**Comment by ridgeworks** [comment]
Just to summarize my experience with 8.1.5-2 (the DIRTY one):

1. Limited testing on my MacBook (MacOS 10.13.6), but as far as I can see *all* original issues resolved. 

2. Crashes on startup on my 2009 Mac Pro running same OS version as above in either command line or bundle mode. I would normally suspect something wrong with my config, except that v8.1.4 runs fine. It may well be a local issue, but I don't know how to determine that. (This is a bit of a showstopper for me, so I need to sort out the cause moving forward.)

My opinion: If Qt 5.12 is required and it, in turn, requires MacOS 10.12, so be it. We all have to move on sometime and there are the alternatives you listed, as well as the 8.1.4 bundle configuration with minor missing functionality (menus). And I wouldn't even mind foregoing the native Font menu if the default was fixed width. (Not sure what the Colors menu even does.)

**Comment by JanWielemaker** [comment]
Thanks.  The Mac Pro issue is worrying.  Possibly `otools` can help figuring out the wrong binding.   8.1.4 is built on 10.11.   It might be ok to support only 10.12 and up, but failing on a 10.13 version is not good.  Possibly the way I specified the deployment version is not ok, though it surely did something ...

**Comment by ridgeworks** [comment]
I do have otool installed on the Mac Pro as part of the command line developer package, but no real idea of what to do with it.

Whatever you did specifying the development version definitely worked on the MacBook, and at least permitted launching the bundle on the Mac Pro.

**Comment by JanWielemaker** [comment]
`otools -L <program or dylib>` tells you to which other dylibs the file binds.  Binding to a different (system) library may give a hint.

**Comment by JanWielemaker** [comment]
According to https://stackoverflow.com/questions/17461196/determining-os-x-sdk-and-deployment-target-versions-from-framework, we can find the minimal deployment version for .dylibs and executables using `otool -l` and looking for `LC_VERSION_MIN_MACOSX`.  It turns out these are not set on the dependencies that have been built with Macports and are included in the bundle.  That _might_ be an explanation.  I'll try to figure out how to reinstall all dependencies with the min version target set.

**Comment by ridgeworks** [comment]
In parallel I'm running otool on the Frameworks directory contents on both the MacPro and MacBook. So far they look identical (still need to check the QT libraries). Let me know if this is useful information.

**Comment by ridgeworks** [comment]
I executed:
`otool -L /Applications/SWI-Prolog-2.app/Contents/Frameworks/*.dylib`

on the MacBook and MacPro and came up with identical results: see attached:

[otoolMacPro.txt](https://github.com/SWI-Prolog/issues/files/3112850/otoolMacPro.txt)

The remaining content in the Frameworks directory are the QT libraries. Running `otool` on the individual libraries gives identical results as well:

[otoolMacProQT.txt](https://github.com/SWI-Prolog/issues/files/3112980/otoolMacProQT.txt)

Also checked the the PlugIns directory with same result:

[otoolMacProPlugIns.txt](https://github.com/SWI-Prolog/issues/files/3113069/otoolMacProPlugIns.txt)

So I can't find any differences between the two systems with respect to libraries required by SWIP.

Did you spot anything in the crash report that might help?

**Comment by JanWielemaker** [comment]
My working hypothesis is that the Macports libraries are the cause.  I have little clue why it works on one and not on the other though.

**Comment by JanWielemaker** [comment]
Uploaded http://eu.swi-prolog.org/download/devel/bin/swipl-8.1.5-3.x86_64.dmg
This version is built against only self compiled dependencies that have their min deployment target set to 10.12.  Please test.

**Comment by ridgeworks** [comment]
Good progress, but not quite there.

Again, no problems (limited testing) on the MacBook.

On the MacPro, the application now launches and runs fine - that's the progress.

However the query:
`?- X is 1*(1 div 2).`
crashes on both the bundle and the command line app. Trace from the terminal:
```
$ /Applications/SWI-Prolog-3.app/Contents/MacOS/swipl 
Welcome to SWI-Prolog (threaded, 64 bits, version 8.1.5-5-g9afb9384b-DIRTY)
SWI-Prolog comes with ABSOLUTELY NO WARRANTY. This is free software.
Please run ?- license. for legal details.

For online help and background, visit http://www.swi-prolog.org
For built-in help, use ?- help(Topic). or ?- apropos(Word).

?- X is 1 rdiv 2.
X = 1 rdiv 2.

?- X is 1*(1 rdiv 2).

SWI-Prolog [thread 1 (main) at Thu Apr 25 17:29:30 2019]: received fatal signal 4 (ill)
C-stack trace labeled "crash":
  [0] /Applications/SWI-Prolog-3.app/Contents/Frameworks/libswipl.8.1.5.dylib(save_backtrace+0xee) [0x107c9094e]
  [1] /Applications/SWI-Prolog-3.app/Contents/Frameworks/libswipl.8.1.5.dylib(sigCrashHandler+0xb2) [0x107c90f72]
  [2] /Applications/SWI-Prolog-3.app/Contents/Frameworks/libswipl.8.1.5.dylib(dispatch_signal+0x233) [0x107c1d1c3]
  [3] /usr/lib/system/libsystem_platform.dylib(_sigtramp+0x1a) [0x7fff64fe6f5a]
  [5] /Applications/SWI-Prolog-3.app/Contents/Frameworks/libgmp.10.dylib(__gmpq_mul+0x1a9) [0x107d16239]
  [6] /Applications/SWI-Prolog-3.app/Contents/Frameworks/libswipl.8.1.5.dylib(ar_mul+0xb0) [0x107bba7c0]
  [7] /Applications/SWI-Prolog-3.app/Contents/Frameworks/libswipl.8.1.5.dylib(valueExpression+0x541) [0x107bb96d1]
  [8] /Applications/SWI-Prolog-3.app/Contents/Frameworks/libswipl.8.1.5.dylib(pl_is2_va+0x67) [0x107bbae47]
  [9] /Applications/SWI-Prolog-3.app/Contents/Frameworks/libswipl.8.1.5.dylib(PL_next_solution+0xdb57) [0x107ba1867]
  [10] /Applications/SWI-Prolog-3.app/Contents/Frameworks/libswipl.8.1.5.dylib(query_loop+0x274) [0x107bf9d14]
  [11] /Applications/SWI-Prolog-3.app/Contents/Frameworks/libswipl.8.1.5.dylib(prologToplevel+0x51) [0x107bfa3d1]
  [12] /Applications/SWI-Prolog-3.app/Contents/Frameworks/libswipl.8.1.5.dylib(PL_toplevel+0x38) [0x107bb32f8]
  [13] /Applications/SWI-Prolog-3.app/Contents/MacOS/swipl(main+0x25) [0x107b85f65]
  [14] /usr/lib/system/libdyld.dylib(start+0x1) [0x7fff64cd8015]
Running on_halt hooks with status 132
Killing 15052 with default signal handlers
Illegal instruction: 4
```

**Comment by JanWielemaker** [comment]
The new machine still works?  I thought this would as simple as the gmp build ignoring the CFLAGS that specifies the OS version, but that is not the case 😞.   Now I'm a little at a loss.   I'm wondering whether the probably older CPU is the issue, but it seems to crash In the system library, possibly as GMP calls one of the memory allocation routines, but it might also be that the actual crash is in GMP and the system library is just the signal handler.

Can you find the CPU for both machines?

**Comment by JanWielemaker** [comment]
Take 4 :smile: http://eu.swi-prolog.org/download/devel/bin/swipl-8.1.5-4.x86_64.dmg
This is with GMP compiled using `--enable-fat` following the suggestion by the lead GMP developer Torbjörn Granlund.   Passes all tests on the build machine.  Please try.

**Comment by ridgeworks** [comment]
All now seems well on the MacPro. I'll do some more testing on both computers and let you know of any issues.

Is this due to some incompatibility at the hardware CPU level?

Thanks for sticking with this; it hasn't been an easy problem to resolve.

**Comment by JanWielemaker** [comment]
Great!   Yes, GMP by default compiles for the native CPU, mine is newer than yours and apparently has some new instructions GMP likes using.   You can tell it to compile for an old CPU, but on some platforms it is possible to have multiple versions of a function and the linker will bind the right one depending on the actual hardware.  That at least is my understanding and ``--enable-fat`` exploits this.

Some more testing might be nice.   @erlanger added an option to add the test suite to the installed system.   Might be an option to try that out.  Almost time to close this :smile:

**Comment by ridgeworks** [comment]
Certainly not my call (and probably too late anyway) but it might better if `--enable-fat` was the default (most general) and leave any fine tuning as a app developer decision. 

Is it possible for me to run the SWIP test suite? If so can you point me at some doc that tells me how?

I did run the installation check (attached). A couple of FAILS but nothing that is of immediate concern to me.

[InstallChk8.1.5-8-gcc5023a8c.txt](https://github.com/SWI-Prolog/issues/files/3121987/InstallChk8.1.5-8-gcc5023a8c.txt)

**Comment by erlanger** [comment]
To run the installed test suite just do this (from within swi-prolog):

```
?- test_installation.
```
It will run all the rests.

If it complains you may need to recompile with `cmake -DINSTALL_TESTS=ON`
